homeassistant/components/enphase_envoy/services.py

```diff
@@ -0,0 +1,92 @@
+"""Implement Enphase Envoy services."""
+
+from __future__ import annotations
+
+import logging
+
+from pyenphase import Envoy
+from pyenphase.const import URL_TARIFF
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntryState
+from homeassistant.core import (
+    HomeAssistant,
+    ServiceCall,
+    ServiceResponse,
+    SupportsResponse,
+)
+from homeassistant.exceptions import ServiceValidationError
+
+from .const import DOMAIN
+from .coordinator import EnphaseUpdateCoordinator
+
+ATTR_CONFIG_ENTRY_ID = "config_entry_id"
+SERVICE_GET_RAW_SCHEMA = vol.Schema(
+    {
+        vol.Required(ATTR_CONFIG_ENTRY_ID): str,
+    }
+)
+RAW_SERVICE_LIST: dict[str, str] = {"get_raw_tariff": URL_TARIFF}
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _find_envoy_coordinator(
+    hass: HomeAssistant, call: ServiceCall
+) -> EnphaseUpdateCoordinator:
+    """Find envoy config entry from service data and return envoy coordinator."""
+    identifier = str(call.data.get(ATTR_CONFIG_ENTRY_ID))
+    if not (entry := hass.config_entries.async_get_entry(identifier)):
+        raise ServiceValidationError(
+            translation_domain=DOMAIN,
+            translation_key="envoy_service_envoy_not_found",
+            translation_placeholders={
+                "service": call.service,
+                "args": identifier,
+            },
+        )
+    if entry.state is not ConfigEntryState.LOADED:
+        raise ServiceValidationError(
+            translation_domain=DOMAIN,
+            translation_key="not_initialized",
+            translation_placeholders={"service": call.service},
+        )
+    coordinator: EnphaseUpdateCoordinator = entry.runtime_data
+    return coordinator
+
+
+async def setup_hass_services(hass: HomeAssistant) -> ServiceResponse:
+    """Configure Home Assistant services for Enphase_Envoy."""
+
+    if hass.services.async_services_for_domain(DOMAIN):
+        return None
+
+    async def get_raw_service(call: ServiceCall) -> ServiceResponse:
+        """Return tariff data from envoy.data.raw cache."""
+        coordinator: EnphaseUpdateCoordinator = _find_envoy_coordinator(hass, call)
+        envoy_to_use: Envoy = coordinator.envoy
```

Comment:
```suggestion
        coordinator = _find_envoy_coordinator(hass, call)
        envoy_to_use = coordinator.envoy
```
I am sure we can remove the coordinator type, not sure about the envoy one

Reply:
Both seem to work in local pre-commits. now pending CI tests.